### PR TITLE
[12.0][IMP] attachment_syncronize, check for duplicated filename

### PR DESCRIPTION
**Pull request by GSLabIt**
- check for duplicates
- prefer enabled in place of active

### CODE REVIEW

File: attachment_synchronize/models/task.py [hunk 1]
```diff
@@ -77,8 +77,9 @@ class StorageTask(models.Model):
         string="File Type",
         help="The file type determines an import method to be used "
              "to parse and transform data before their import in ERP")
-    active = fields.Boolean(default=True)
+    enabled = fields.Boolean('Enabled', default=True)
```
Comment (florian-dacosta) [inline attachment_synchronize/models/task.py:80]: Why this change?
It is not good. The active field is very specific field, present all around Odoo. If active is not checked, the record is invisible in Odoo and ignored by default in the search, etc...
This behavior won't work with a field named "enable."

File: attachment_synchronize/models/task.py [hunk 2]
```diff
@@ -77,8 +77,9 @@ class StorageTask(models.Model):
...
```
Comment (florian-dacosta) [inline attachment_synchronize/models/task.py:84]: useless line break

File: attachment_synchronize/models/task.py [hunk 3]
```diff
@@ -77,8 +77,9 @@ class StorageTask(models.Model):
...
 
+    check_duplicated_files = fields.Boolean('Check duplicated files')
```
Comment (florian-dacosta) [inline attachment_synchronize/models/task.py]: Add a help attribute to explain how the check is done. And warn that it will prevent the import of such files

File: attachment_synchronize/models/task.py [hunk 4]
```diff
@@ -165,3 +170,8 @@ def run_import(self):
                         raise e
                     else:
                         new_env.cr.commit()
+
+    def _duplicated_file(self, filename):
+        if self.attachment_ids.search([('filename', '=', filename)]):
```
Comment (florian-dacosta) [inline attachment_synchronize/models/task.py]: Doing this, you read all attachments linked to the task, there is no need.
Replace by self.env['ir.attachment.metadata'].search()
Add task_id on the domain if you want to filter on attachment linked to this task (make sens to me)
Also, the field filename does not seem to exist. Maybe use the field datas_fname, we are sure the filename is inside : https://github.com/akretion/server-tools/blob/12-mig-external_file_location/attachment_synchronize/models/task.py#L88
You should probably add an index on the field in that case, for performance.

Also, as the system has to loop on all file names from the external system each time the cron runs, it may cause performance issue in the long run.
If your external app never delete or rename the files, you will end having issues I think

**Review by florian-dacosta — CHANGES_REQUESTED** [review]

File: attachment_synchronize/models/task.py (hunk 1 above)
Comment (GSLabIt) [reply attachment_synchronize/models/task.py:80]: because it will disappear from tasks tree view in storage_backend, and if deactivated in error it will not be shown anymore, that's the only reason. 
That said no problem to revert as active.

File: attachment_synchronize/models/task.py (hunk 4 above)
Comment (GSLabIt) [reply attachment_synchronize/models/task.py]: well that's should not reading to all attachment, but searching on all attachment related to the task; probably filtered will be a better solution in this case, so kind of:
```
    if self.attachment_ids.filtered(lambda r: r.name == filename)
```
you're right, i meant to search for `name`. 

Yes, looping will be expensive in the long run, but actually it's better than process same file multiple times. 

I'll need to develop storage_backend_ftp, probably i could implement some kind of logic to list file modified later than last sync or similar.

File: attachment_synchronize/models/task.py (hunk 1 above)
Comment (florian-dacosta) [reply attachment_synchronize/models/task.py:80]: I don't get it, disapear from task tree view? `attachment_ids` is not present in task view so the attachment should never appear in task view.
If unactive by error, it can be active again to correct the error
I mean, the purpose of active field is to archive stuff and so make it disappear, I think it is good to has it yes

Comment (GSLabIt) [reply attachment_synchronize/models/task.py:80]: ok, I try to explain better. The `active` field is related to the task not to the `attachment_ids`. If u uncheck it, will be archived and it will not be shown in task tree view inside `storage_backend` form view. So, actually i prefer to see task in the tree view and disable it, of course in this case we'll need to check if task enabled once executing cron job.

File: attachment_synchronize/models/task.py (hunk 4 above)
Comment (florian-dacosta) [reply attachment_synchronize/models/task.py]: Well, I am not sure if filtered is a better option.
I mean, it depends if the external storage is sometimes cleaned or not...

Doing the filtered will be very expensive, because it needs to read all attachment linked to the task, which can be huge with time.
But once done, the result is put in cache.

Althoug a search will be really faster, in case you loop on a huge number of file, it will end up costing also...
So it will kind of depend of the use case.

About storage_backend_ftp, you should not need this as you could rename/move file with ftp right?

Anyway, I let you decide about the filtered or the search...

File: attachment_synchronize/models/task.py (hunk 1 above)
Comment (florian-dacosta) [reply attachment_synchronize/models/task.py:80]: Oh yes, my bad, it is on the task, I was thinking of the attachment sorry!
You are right then, if unactive, it can't be active again from the UI, this may be an issue.
well, the enable field seems like a good idea then but such task should be ignored indeed

Sorry for the noise!

File: attachment_synchronize/models/task.py (hunk 4 above)
Comment (GSLabIt) [reply attachment_synchronize/models/task.py]: Sorry i'm not getting the difference by searching using `attach_obj = self.env['ir.attachment.metadata']`in place of `attachment_ids`, could you please let me know. 

Probably would be faster if searching for a list of names in `self.env['ir.attachment.metadata']`and then get list difference and loop on that to import files.

> About storage_backend_ftp, you should not need this as you could rename/move file with ftp right?

What u mean by this?

Comment (florian-dacosta) [reply attachment_synchronize/models/task.py]: `attach_obj = self.env['ir.attachment.metadata']` is just an empty recordset
`while self.attachment_ids` is a recordset of the attachments linked to the task.
So both correspond to the same object, the difference is that the first one is empty and the second one is a recordset (which maybe huge).

So, if you don't actually need to read the linked attachment, it is better to go with an empty recordset.
But if you need to actually read the attachments (in case of filtered), then it is fine.

Anyway, if you do a `search`on a recordset, Odoo behave exatly the same as if you do a search on an empty recordset. Meaning that, if you do self.attachment_ids.search(...) Odoo will search among all attachments, and not only the one from the recordset (the one linked to the task). I am not sure I am very clear...

> Probably would be faster if searching for a list of names in self.env['ir.attachment.metadata']and then get list difference and loop on that to import files.

Probably a good compromise indeed.

`What u mean by this?`
What I said earlier. The module allow you to move/rename a file after importing.
You should be able to do this by ftp (if you have the write rights on the ftp server) in this case, the check duplicated feature is not needed !

Comment (GSLabIt) [reply attachment_synchronize/models/task.py]: As far as i know, but at this point not anymore sure i got in the right way, searching by empty record set will search on all the table kind of:
```
     select * from ir_attachment_metadata
```
while `attachment_ids` should be already in memory object, so actually the only related to the task. I thought searching within this recordset will ended searching in memory instead of making query to the db, and this should be faster and over a smaller recordset. 

But sound like i misunderstood something 😬, that said no prob to use empty recordset if this is faster.

> What I said earlier. The module allow you to move/rename a file after importing.
> You should be able to do this by ftp (if you have the write rights on the ftp server) in this case, the check duplicated feature is not needed !

Yup, but as i was telling you, i can't delete or move files because other apps, beside mine, will have to download and process.

Comment (florian-dacosta) [reply attachment_synchronize/models/task.py]: ` I thought searching within this recordset will ended searching in memory instead of making query to the db, and this should be faster and over a smaller recordset.`

No, a search will always do queries to the database.

File: attachment_synchronize/models/task.py [hunk 5]
```diff
@@ -126,6 +129,8 @@ def run_import(self):
         backend = self.backend_id
         filenames = backend._list(
             relative_path=self.filepath, pattern=self.pattern)
+        if self.check_duplicated_files:
+            filenames = self._file_to_import(filenames)
```
Comment (GSLabIt) [inline attachment_synchronize/models/task.py:133]: @florian-dacosta isn't `filenames` returned by `backand._list` a list of names? am I right?

**Comment by GSLabIt** [comment]
@florian-dacosta after further thinking, i've decided to go using `filtered` as it will work on in memory recordset.

File: attachment_synchronize/models/task.py (hunk 5 above)
Comment (florian-dacosta) [reply attachment_synchronize/models/task.py:133]: AFAIK it is, yes

**Review by florian-dacosta — APPROVED** [review]
Seems good to me.
I did not test anything though.
We'll need to add some tests. (the test inside the module are obsolete and were for the old version)
